heft-plugins/heft-jest-plugin/src/JestPlugin.ts

```diff
@@ -97,60 +101,152 @@ const JEST_CONFIG_PACKAGE_FOLDER: string = path.dirname(require.resolve('jest-co
+    ) as CommandLineStringParameter;
+    const testTimeout: CommandLineStringParameter = taskSession.parametersByLongName.get(
+      '--test-timeout-ms'
+    ) as CommandLineStringParameter;
+    const findRelatedTests: CommandLineStringParameter = taskSession.parametersByLongName.get(
+      '--find-related-tests'
+    ) as CommandLineStringParameter;
+    const testNamePattern: CommandLineStringParameter = taskSession.parametersByLongName.get(
+      '--test-name-pattern'
+    ) as CommandLineStringParameter;
+    const testPathPattern: CommandLineStringParameter = taskSession.parametersByLongName.get(
+      '--test-path-pattern'
+    ) as CommandLineStringParameter;
+
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Jest's cache is not reliable.  For example, if a Jest configuration change causes files to be
+      // transformed differently, the cache will continue to return the old results unless we manually
+      // clean it.  Thus we need to ensure that we always cleans the Jest cache.
+      cleanOptions.addDeleteOperations({ sourceFolder: taskSession.cacheFolder });
+
+      // We should also clean the data file that we generate for the BuildTransformer
+      const dataFilePath: string = HeftJestDataFile.getConfigFilePath(heftConfiguration.buildFolder);
+      cleanOptions.addDeleteOperations({
+        sourceFolder: path.dirname(dataFilePath),
+        includeGlobs: [path.basename(dataFilePath)]
+      });
+    });
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      const combinedOptions: IJestPluginOptions = {
+        ...pluginOptions,
+        configurationPath: config.value || pluginOptions?.configurationPath,
+        debugHeftReporter: debugHeftReporter.value || pluginOptions?.debugHeftReporter,
+        detectOpenHandles: detectOpenHandles.value || pluginOptions?.detectOpenHandles,
+        disableCodeCoverage: disableCodeCoverage.value || pluginOptions?.disableCodeCoverage,
+        findRelatedTests: findRelatedTests.value || pluginOptions?.findRelatedTests,
+        maxWorkers: maxWorkers.value || pluginOptions?.maxWorkers,
+        // Default to true and always pass with no tests
+        passWithNoTests: true,
+        silent: silent.value || pluginOptions?.silent,
+        testNamePattern: testNamePattern.value || pluginOptions?.testNamePattern,
+        testPathPattern: testPathPattern.value || pluginOptions?.testPathPattern,
+        testTimeout: testTimeout.value ? parseInt(testTimeout.value, 10) : pluginOptions?.testTimeout,
+        updateSnapshots: updateSnapshots.value || pluginOptions?.updateSnapshots
+      };
+      await this._runJestAsync(taskSession, heftConfiguration, combinedOptions);
+    });
+  }
+
+  /**
+   * Write the data file used by the BuildTransformer
    */
-  public static async _setupJestAsync(
-    scopedLogger: ScopedLogger,
+  private async _setupJestAsync(
```

Comment:
Wasn't this previously the responsibility of the typescript plugin?

Reply:
Part of having TS write this file was that the Jest plugin could ensure that the project had been built before we attempted to run tests.

Reply:
It wasn't the responsibility of the typescript plugin, _but_ the typescript plugin provided the data to the stage, which the Jest plugin later picked up on and used. In new Heft, the `test` phase will have no insight into the `build` phase, and is self-contained, so we need to separately load it up (if present) to determine this information. If we were to continue doing it the old way, we'd need to have a separate Jest setup plugin which runs during the build phase, and then use the test phase to pick up the resulting file. I originally tried this, but it just felt... wrong. There should be no need to do that when we can load in the data entirely separately, without requiring the use of 2 separate plugins.

A nice side effect is that this also allows us to be able to use the Jest plugin whether or not TypeScript is used in a Heft project.